src/hooks/useTransactions.tsx

```diff
@@ -1,73 +1,50 @@
 import React from "react";
 import { ethers } from "ethers";
+import { useQueries, useQueryClient, UseQueryResult } from "react-query";
 
 type Transaction = ethers.providers.TransactionResponse;
-type TransactionReceipt = ethers.providers.TransactionRequest;
-type TransactionStatus = "idle" | "pending" | "resolved" | "rejected";
+type TransactionReceipt = ethers.providers.TransactionReceipt;
 
 type TransactionState = {
-  transaction?: Transaction;
-  status?: TransactionStatus;
+  transactions: Transaction[];
+  transactionQueries?: UseQueryResult<TransactionReceipt>[];
   error?: Error;
 };
 
-enum ActionType {
-  ERROR,
-  UPDATE,
-}
-
-type Action =
-  | {
-      type: ActionType.UPDATE;
-      payload: TransactionState;
-    }
-  | {
-      type: ActionType.ERROR;
-      payload: Pick<TransactionState, "error">;
-    };
-function reducer(state: TransactionState, action: Action) {
-  switch (action.type) {
-    case ActionType.UPDATE: {
-      return {
-        ...state,
-        ...action.payload,
-      };
-    }
-    case ActionType.ERROR: {
-      const { error } = action.payload;
-      return {
-        ...state,
-        error,
-      };
-    }
-  }
-}
-
 type TransactionsManagerState = {
   addTransaction: (
     tx: Transaction,
     onComplete?: (tx: TransactionReceipt) => void
   ) => void;
 } & TransactionState;
-function useTransactionsManager(): TransactionsManagerState {
-  const [state, dispatch] = React.useReducer(reducer, {});
-  const { transaction, status, error } = state;
+function useTransactionsManager() {
+  const [transactions, setTransactions] = React.useState<Transaction[]>([]);
+  const [error, setError] = React.useState<Error>();
+
+  const transactionQueries = useQueries(
+    transactions.map((tx) => ({
+      queryFn: () => tx.wait(),
```

Comment:
I think ethers will return a new Promise each time `wait` is called, but even if it were the same promise, both `.then` attached to it would be called once the promise resolve and run like normal js functions.
There could be race condition in which one completes first, but I don't think the order in which the attached `.then` resolves matters to our specific usecase